### PR TITLE
Double metadata

### PR DESCRIPTION
**Pull request by yngve793**
DoubleVolumeDataLayout is introduced as a layout class for two VDS files. This class is an implementation of the OpenVDS::VolumeDataLayout abstract class. Introducing an implementation of the layout along with a double variant of DataHandle allows for the reuse of most of the code that interact with OpenVDS. 

CTests and appropriate test data are added. So, metadata, fence, slice, and attribute requests are tested.

I suggest to move go and e2e tests to a new PR.

### CODE REVIEW

**Review by achaikou — COMMENTED** [review]
Nice work!.. :pouting_cat:  I thought I could find the tests that would thoroughly break the feature, but I couldn't... Where is the fun in it? :pouting_cat: :smile: 

This is a big PR, but I agree that it would be difficult to split it more as one has to make sure that everything works. Other division would have been to implement all the stuff locally and to send for PR only general code + slice code + all the possible slice tests. Then, after merge, repeat with fence/attributes. But not sure it would have been better or not. I would have likely complained any way :smile:

File: internal/core/volumedatalayout.cpp [hunk 1]
```diff
@@ -0,0 +1,155 @@
...
+        return m_layout_a->GetDimensionName(dimension);
+    } else {
+        throw detail::bad_request("Dimension name mismatch for dimension: " + std::to_string(dimension));
+    }
+}
+
+const char* DoubleVolumeDataLayout::GetDimensionUnit(int dimension) const {
+    const char* unit_a = m_layout_a->GetDimensionUnit(dimension);
+    const char* unit_b = m_layout_b->GetDimensionUnit(dimension);
+    if (strcmp(unit_a, unit_b) == 0) {
```
[the remark this answers is not in the document]
Comment (yngve793) [reply internal/core/volumedatalayout.cpp:134]: All checks are moved to the constructor. 
Signature of the function is defined by OpenVDS and can not be changed by us.

File: internal/core/volumedatalayout.cpp [hunk 2]
```diff
@@ -0,0 +1,155 @@
...
+
+float DoubleVolumeDataLayout::GetDimensionMin(int dimension) const {
+    return m_dimensionCoordinateMin[dimension];
+}
+
+float DoubleVolumeDataLayout::GetDimensionMax(int dimension) const {
+    return m_dimensionCoordinateMax[dimension];
+}
+
+float DoubleVolumeDataLayout::GetDimensionIndexOffset_a(int dimension) const {
```
[the remark this answers is not in the document]
Comment (yngve793) [reply internal/core/volumedatalayout.cpp:149]: No reason for using float, switched to int. 

Offset for cube A to reach the start of (A \cap B).
Doc added.

File: internal/core/volumedatalayout.hpp [hunk 3]
```diff
@@ -0,0 +1,176 @@
+#ifndef VDL_VOLUME_DATA_LAYOUT_HPP
```
[the remark this answers is not in the document]
Comment (yngve793) [reply internal/core/volumedatalayout.hpp:1]: Ok, fixed.

File: internal/core/volumedatalayout.hpp [hunk 4]
```diff
@@ -1,5 +1,5 @@
-#ifndef VDL_VOLUME_DATA_LAYOUT_HPP
-#define VDL_VOLUME_DATA_LAYOUT_HPP
+#ifndef VDL_SLICE_VOLUME_DATA_LAYOUT_HPP
```
Comment (achaikou) [inline internal/core/volumedatalayout.hpp:1]: VD**S**_SLICE x3

File: internal/core/datahandle.cpp [hunk 5]
```diff
@@ -269,23 +290,85 @@ void DoubleDataHandle::read_traces(
...
+        int index = i % traceIndex_a;
+
+        if (index >= coordinates_a[dimension] && index < coordinates_a[dimension] + nsamples) {
+            floatBuffer[counter] = buffer_a[i];
+            counter++;
+        }
+    }
+
+    counter = 0;
+    int traceIndex_b = this->m_metadata_b.sample().nsamples();
```
[the remark this answers is not in the document]
Comment (yngve793) [reply internal/core/datahandle.cpp:372]: Created new function 'extract_part_of_trace' with better naming.

**Review by yngve793 — COMMENTED** [review]
Continue fixing and replaying to comments.

**Review by achaikou — COMMENTED** [review]
Not sure about correctness of choice of comments for some fixups, but lets leave that possible git rebase hell for later. :skull_and_crossbones:

**Review by yngve793 — COMMENTED** [review]
Most comments are now addressed. I think that it can be a good point to have a new iteration. For each point it is a trade off between cleaning it up completely or checking if it needs to be rewritten. 

Main changes are removal of datasource and setting gtests for the different requests up in a system. I think that it is wise to agree on the system and structure before continuing. 

Github simply links the comment to a file and not a commit. This makes it hard to connect the comment to the right commit and context. 

Refactoring has caused some of the fixups to be merged naturally while other very small ones are merged to reduce the number of commits. I hope that the tradeoff here makes it easier to review.  It is also a bit unclear what will be removed or rewritten in the coming restructuring.

File: testdata/cube_intersection/convert_intersection.sh
Comment (achaikou) [inline testdata/cube_intersection/convert_intersection.sh:1]: Btw, why did you need these files in Azure? Do you think we would need to test these features with E2E too?
I don't see any need currently.

File: cmd/query/main_test.go [hunk 6]
```diff
@@ -719,7 +719,7 @@ func TestMetadataErrorHTTPResponse(t *testing.T) {
 				name:           "Two urls to metadata request",
 				method:         http.MethodPost,
 				expectedStatus: http.StatusBadRequest,
-				expectedError:  "Metadata requests only accepts one VDS url and one sas token",
+				expectedError:  "Binary operator must be provided when two VDS urls are provided",
```
[the remark this answers is not in the document]
Comment (achaikou) [reply cmd/query/main_test.go:722]: As agreed: probably rename the test and create two in total - one with binary operator another without.

**Review by achaikou — COMMENTED** [review]
Nice tests! Me likes that! :cat: 

Closed all the comments that will be dealt with after merge.
Added some new ones, hopefully nothing groundbreaking. Some comments repeated for better tracking.

Go through remaining comments from previous iterations, if they are not resolved, I want something there still, mostly discussed in person. Hopefully you remember what, as these comments are no longer trackable due to fixups merge. If not, just ask.

**Review by yngve793 — COMMENTED** [review]
All comments that are listed in commits are addressed, but it is hard to follow when some files have changed and some commits are squashed. 

During the process of going through the comments several inconsistencies was discovered. White char at end of line, etc. To reduce these to a minimum 'pre_commit' tests was added. As a result of this the first 5 commits are dedicated to getting existing files up to standard. 

All fixups from previous review are applied to the existing commits. I think that all but one of the new fixups for the comments are i separate fixups. I have added hash code for some of the fixups, but are uncertain if some of them changed in the final rebase.

File: tests/gtest/datahandle_metadata_test.cpp [hunk 7]
```diff
@@ -161,4 +163,150 @@ TEST_F(DatahandleMetadataTest, Metadata_Different_Size_Double) {
     EXPECT_EQ(metadata["axis"], different_size_metadata["axis"]);
 }
 
+TEST_F(DatahandleMetadataTest, Metadata_One_crossLine_Single) {
+
+    // const std::string SINGLE_XLINE_DATA = "file://10_single_xline.vds";
```
Comment (achaikou) [inline tests/gtest/datahandle_metadata_test.cpp:168]: :broom:  ? Or  on the opposite, remove it from the named files to make consistent with tests below?

File: internal/core/datahandle.cpp [hunk 8]
```diff
@@ -207,6 +207,10 @@ DoubleDataHandle::DoubleDataHandle(OpenVDS::VDSHandle handle_a, OpenVDS::VDSHand
         this->m_binary_operator = &inplace_division;
     else
         throw detail::bad_request("Invalid binary_operator string");
+
+    if (this->m_metadata_a.crs().compare(this->m_metadata_b.crs()) != 0) {
```
[the remark this answers is not in the document]
Comment (achaikou) [reply internal/core/datahandle.cpp:210]: I think you can put it to the doublevolumedatalayout constructor, near the origin comparison and use layouts instead of the metadatas.

But if you are not in the mood, I can let it slip :)

**Review by achaikou — COMMENTED** [review]
Love the new tests and new restrictions! :heart_eyes_cat: 

And I feel like :champagne: should be on its way!

**Review by yngve793 — COMMENTED** [review]
Fixes are in the 5 commits at the end.

**Review by achaikou — COMMENTED** [review]
One comment only.
Please apply all the fixups so that no are left (including the new one). I will take a final look at this PR as the whole, to make sure we didn't mess something up majorly during rebasing, and then hopefully :champagne:

**Comment by yngve793** [comment]
Sorry about the slip up. I lost my changes when rebasing and had to stat over. 

All commits are compressed and organised as much as possible.
